search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,433 @@
+
+static void
+add_key_value_pair_to_variant (GVariantBuilder *builder,
+                               const char      *key,
+                               GVariant        *value)
+{
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("{sv}"));
+  g_variant_builder_add (builder, "s", key);
+  g_variant_builder_add (builder, "v", value);
+  g_variant_builder_close (builder);
+}
+
+static inline gchar *
+underscorify (const gchar *string)
+{
+  gchar *underscored = strdup (string);
+  gchar *c = underscored;
+  do
+    {
+      if (*c == '-')
+        *c = '_';
+    }
+  while (*c++);
+
+  return underscored;
+}
+
+static void
+maybe_add_key_value_pair_from_model_to_variant (EkncContentObjectModel *model,
+                                                GVariantBuilder        *builder,
+                                                const char             *key,
+                                                const GVariantType     *expected_type)
+{
+  g_auto(GValue) value = G_VALUE_INIT;
+  g_autofree gchar *underscore_key = underscorify (key);
+  GParamSpec *pspec = g_object_class_find_property (G_OBJECT_GET_CLASS (model),
+                                                    key);
+
+  if (pspec == NULL)
+    return;
+
+  g_value_init (&value, pspec->value_type);
+  g_object_get_property (G_OBJECT (model), key, &value);
+  add_key_value_pair_to_variant (builder,
+                                 underscore_key,
+                                 g_dbus_gvalue_to_gvariant (&value,
+                                                            expected_type));
+}
+
+static void
+on_received_query_results (GObject      *source,
+                           GAsyncResult *result,
+                           gpointer      user_data)
+{
+  EkncEngine *engine = EKNC_ENGINE (source);
+  g_autoptr(ArticleMetadataQueryState) state = user_data;
+  g_autoptr(ObjectSList) models = NULL;
+  g_autoptr(ObjectSList) shards = NULL;
+  g_auto(GStrv) shards_strv = NULL;
+  g_auto(GVariantBuilder) builder;
+  g_autoptr(GError) error = NULL;
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("aa{sv}"));
+
+  g_application_release (g_application_get_default ());
+
+  if (!models_and_shards_for_result (engine,
+                                     state->provider->application_id,
+                                     result,
+                                     (GSList **) &models,
+                                     (GSList **) &shards,
+                                     &error))
+    {
+      g_dbus_method_invocation_take_error (state->invocation,
+                                           g_steal_pointer (&error));
+      return;
+    }
+
+  shards_strv = strv_from_shard_list ((GSList *) shards);
+
+  for (GSList *l = (GSList *) models; l; l = l->next)
+    {
+      EkncContentObjectModel *model = l->data;
+
+      g_variant_builder_open (&builder, G_VARIANT_TYPE("a{sv}"));
+
+      maybe_add_key_value_pair_from_model_to_variant (model,
```

Comment:
It's a bit awkward to do this GValue->GVariant dance here, mostly because we IIRC do the inverse in the eknc code itself; but I guess it's OK.